pkg/apis/apps/types.go

```diff
@@ -60,6 +60,50 @@ const (
 	ParallelPodManagement = "Parallel"
 )
 
+// StatefulSetUpdateStrategy indicates the strategy that the StatefulSet
+// controller will use to perform updates. It includes any additional parameters
+// necessary to preform the update for the indicated strategy.
+type StatefulSetUpdateStrategy struct {
+	// Type indicates the type of the StatefulSetUpdateStrategy.
+	Type StatefulSetUpdateStrategyType
+	// Partition is used to communicate the ordinal at which to partition
+	// the StatefulSet when Type is PartitionStatefulSetStrategyType. This
+	// value must be set when Type is PartitionStatefulSetStrategyType,
+	// and it must be nil otherwise.
+	Partition *PartitionStatefulSetStrategy
+}
+
+// StatefulSetUpdateStrategyType is a string enumeration type that enumerates
+// all possible update strategies for the StatefulSet controller.
+type StatefulSetUpdateStrategyType string
+
+const (
+	// PartitionStatefulSetStrategyType indicates that updates will only be
+	// applied to a partition of the StatefulSet. This is useful for canaries
+	// and phased roll outs. When a scale operation is performed with this
+	// strategy, new Pods will be created from the updated specification.
+	PartitionStatefulSetStrategyType StatefulSetUpdateStrategyType = "Partition"
+	// RollingUpdateStatefulSetStrategyType indicates that update will be
+	// applied to all Pods in the StatefulSet with respect to the StatefulSet
+	// ordering constraints. When a scale operation is performed with this
+	// strategy, new Pods will be created from the updated specification.
+	RollingUpdateStatefulSetStrategyType = "RollingUpdate"
+	// OnDeleteStatefulSetStrategyType triggers the legacy behavior. Version
+	// tracking and ordered rolling restarts are disabled. Pods are recreated
+	// from the StatefulSetSpec when they are manually deleted. When a scale
+	// operation is performed with this strategy, new Pods will be created
+	// from the current specification.
+	OnDeleteStatefulSetStrategyType = "OnDelete"
+)
+
+// PartitionStatefulSetStrategy contains the parameters used with the
+// PartitionStatefulSetStrategyType.
+type PartitionStatefulSetStrategy struct {
+	// Ordinal indicates the ordinal at which the StatefulSet should be
+	// partitioned.
+	Ordinal int32
```

Comment:
I think we should require partition to be specified.  The default doesn't add anything for the machine operated use case which is going to be much more common